rmgpy/solver/base.pyx

```diff
@@ -218,11 +232,76 @@ cdef class ReactionSystem(DASx):
         self.sensitivityCoefficients = numpy.zeros((self.numCoreSpecies, self.numCoreReactions), numpy.float64)
         self.unimolecularThreshold = numpy.zeros((self.numCoreSpecies), bool)
         self.bimolecularThreshold = numpy.zeros((self.numCoreSpecies, self.numCoreSpecies), bool)
-        
 
+        surfaceSpecies,surfaceReactions = self.initialize_surface(coreSpecies,coreReactions,surfaceSpecies,surfaceReactions)
+        
+        
     def initialize_solver(self):
         DASx.initialize(self, self.t0, self.y0, self.dydt0, self.senpar, self.atol_array, self.rtol_array)
-
+    
+    @cython.boundscheck(False)
+    cpdef initialize_surface(self,list coreSpecies,list coreReactions,list surfaceSpecies,list surfaceReactions):
+        """
+        removes surfaceSpecies and surfaceReactions from  until they are self consistent: 
+            1) every reaction has one species in the surface
+            2) every species participates in a surface reaction
+        """
+        cdef numpy.ndarray[numpy.int_t,ndim=2] productIndices,reactantIndices
+        cdef list surfaceSpeciesIndices, surfaceReationIndices
+        cdef set possibleSpeciesIndices
+        cdef int i,j
+        cdef bool notInSurface
+        
+        logging.info('initializing surface ...')
+        
+        productIndices = self.productIndices
+        reactantIndices = self.reactantIndices
+        
+        surfaceSpeciesIndices = []
+        surfaceReactionIndices = []
+        possibleSpeciesIndices = set()
+        
+        for i in range(len(surfaceSpecies)):
```

Comment:
Is the loop style
`for i in range(len(someList)):  do_something_to(someList[i])`
rather than
`for someItem in someList: do_something_to(someItem)`
a cython optimization, or just a habit?